src/http.rs

```diff
@@ -45,77 +46,151 @@ async fn handle_web_req(
-            let mut response = Response::new(Body::from(uri.clone()));
-            create_qr_code(&uri, &address.to_string());
-            response
-                .headers_mut()
-                .insert(hyper::header::CONTENT_TYPE, "text/plain".parse().unwrap());
-            Ok(response)
-        }
+async fn handle_pj(scheduler: Scheduler, req: Request<Body>) -> Result<Response<Body>, HttpError> {
+    dbg!(req.uri().query());
 
-        // Return the 404 Not Found for other routes.
-        _ => {
-            let mut not_found = Response::default();
-            *not_found.status_mut() = StatusCode::NOT_FOUND;
-            Ok(not_found)
+    let headers = Headers(req.headers().to_owned());
+    let query = {
+        let uri = req.uri();
+        if let Some(query) = uri.query() {
+            Some(&query.to_owned());
         }
-    }
+        None
+    };
+    let body = req.into_body();
+    let bytes = hyper::body::to_bytes(body).await?;
+    dbg!(&bytes); // this is correct by my accounts
+    let reader = &*bytes;
+    let original_request = UncheckedProposal::from_request(reader, query, headers)?;
+
+    let proposal_psbt = scheduler.propose_payjoin(original_request).await?;
+
+    Ok(Response::new(Body::from(proposal_psbt)))
+}
+
+async fn handle_pj_schedule(
+    scheduler: Scheduler,
+    endpoint: url::Url,
+    req: Request<Body>,
+) -> Result<Response<Body>, HttpError> {
+    let bytes = hyper::body::to_bytes(req.into_body()).await?;
+    // deserialize x-www-form-urlencoded data with non-strict encoded "channel[arrayindex]"
+    let conf = serde_qs::Config::new(2, false);
+    let request: ScheduledPayJoin = conf.deserialize_bytes(&bytes)?;
+
+    let address = scheduler.schedule_payjoin(&request).await?;
+    let total_amount = request.total_amount();
+    let uri = scheduler::format_bip21(address.clone(), total_amount, endpoint);
+    let mut response = Response::new(Body::from(uri.clone()));
+    create_qr_code(&uri, &address.to_string());
+    response.headers_mut().insert(hyper::header::CONTENT_TYPE, "text/plain".parse()?);
+    Ok(response)
 }
 
 pub(crate) struct Headers(hyper::HeaderMap);
 impl bip78::receiver::Headers for Headers {
     fn get_header(&self, key: &str) -> Option<&str> { self.0.get(key)?.to_str().ok() }
 }
+
+#[derive(Debug)]
+pub enum HttpError {
+    Bip78Request(bip78::receiver::RequestError),
+    Hyper(hyper::Error),
+    Http(hyper::http::Error),
+    InvalidHeaderValue(hyper::header::InvalidHeaderValue),
+    Scheduler(SchedulerError),
+    Serde(serde_qs::Error),
+}
+
+impl HttpError {
+    /// Transforms an [HttpError] into a HTTP response
+    pub fn into_response(self) -> Result<Response<Body>, hyper::Error> {
+        if let Self::Hyper(err) = self {
+            return Err(err);
+        }
+
+        let resp = Response::new(Body::from(self.to_string()));
+        let (mut parts, body) = resp.into_parts();
+
+        // TODO respond with well known errors as defined in BIP-0078
```

Comment:
Will this also give us more informative errors on most senders?

Reply:
The only difference to start would be UI would say

"not-enough-money | The receiver added some inputs but could not bump the fee of the payjoin proposal."

or

"original-psbt-rejected | The receiver rejected the original PSBT. "

otherwise the spec says to put errors into logs only. This becomes more useful when we're using our own sender because we can show can decide to propagate new well-known errors into our UIs.

https://github.com/bitcoin/bips/blob/master/bip-0078.mediawiki#receivers-well-known-errors